File: tests/oracle/printlanguage_group_1204.cc
#include "printc.hh"

#include <iostream>
#include <sstream>
#include <string>

namespace {

using namespace ghidra;

class FixturePrintC final : public PrintC {
public:
  FixturePrintC(void) : PrintC(nullptr, "printlanguage-group-1204") {}

  void rootUnary(void) {
    const std::string x("x");
    pushOp(&boolean_not, nullptr);
    pushAtom(Atom(x, syntax, EmitMarkup::no_color));
    emit->flush();
  }

  void rootBinary(void) {
    const std::string a("a");
    const std::string b("b");
    pushOp(&binary_plus, nullptr);
    pushAtom(Atom(a, syntax, EmitMarkup::no_color));
    pushAtom(Atom(b, syntax, EmitMarkup::no_color));
    emit->flush();
  }

  void nestedInvisible(void) {
    const std::string a("a");
    const std::string b("b");
    const std::string c("c");
    pushOp(&binary_plus, nullptr);
    pushAtom(Atom(a, syntax, EmitMarkup::no_color));
    pushOp(&multiply, nullptr);
    pushAtom(Atom(b, syntax, EmitMarkup::no_color));
    pushAtom(Atom(c, syntax, EmitMarkup::no_color));
    emit->flush();
  }

  void nestedParenthesized(void) {
    const std::string a("a");
    const std::string b("b");
    const std::string c("c");
    pushOp(&multiply, nullptr);
    pushAtom(Atom(a, syntax, EmitMarkup::no_color));
    pushOp(&binary_plus, nullptr);
    pushAtom(Atom(b, syntax, EmitMarkup::no_color));
    pushAtom(Atom(c, syntax, EmitMarkup::no_color));
    emit->flush();
  }
};

template <typename Callback>
std::string render(Callback callback) {
  std::ostringstream output;
  FixturePrintC printer;
  printer.setOutputStream(&output);
  callback(printer);
  return output.str();
}

} // namespace

int main() {
  std::cout << "root_unary="
            << render([](FixturePrintC &printer) { printer.rootUnary(); })
            << '\n';
  std::cout << "root_binary="
            << render([](FixturePrintC &printer) { printer.rootBinary(); })
            << '\n';
  std::cout << "nested_invisible="
            << render([](FixturePrintC &printer) { printer.nestedInvisible(); })
            << '\n';
  std::cout << "nested_parenthesized="
            << render([](FixturePrintC &printer) {
                 printer.nestedParenthesized();
               })
            << '\n';
  return 0;
}

File: tests/oracle/printlanguage_group_1204.rs
use rugra::prettyprint::EmitNoMarkup;
use rugra::printlanguage::{
    rpn_push_atom, rpn_push_op, Atom, NodePending, OpToken, ReversePolish,
    SyntaxHighlight, TagType,
};

fn atom(name: &str) -> Atom {
    Atom::new(name, TagType::Syntax, SyntaxHighlight::VarColor)
}

fn render(callback: impl FnOnce(
    &mut Vec<ReversePolish>,
    &mut Vec<NodePending>,
    &mut usize,
    &[OpToken],
    &mut EmitNoMarkup,
)) -> String {
    let tokens = vec![
        OpToken::unary_prefix("!", 62, 0, 0),
        OpToken::binary("+", 50, true, 1, 0, -1),
        OpToken::binary("*", 54, true, 1, 0, -1),
    ];
    let mut revpol = Vec::new();
    let mut nodepend = Vec::new();
    let mut pending = 0;
    let mut emit = EmitNoMarkup::new();
    callback(
        &mut revpol,
        &mut nodepend,
        &mut pending,
        &tokens,
        &mut emit,
    );
    assert!(revpol.is_empty(), "fixture must completely drain the RPN stack");
    emit.get_output()
}

fn main() {
    let root_unary = render(|revpol, nodepend, pending, tokens, emit| {
        rpn_push_op(revpol, nodepend, pending, tokens, emit, 0, -1);
        rpn_push_atom(revpol, nodepend, pending, tokens, emit, &atom("x"));
    });
    println!("root_unary={root_unary}");

    let root_binary = render(|revpol, nodepend, pending, tokens, emit| {
        rpn_push_op(revpol, nodepend, pending, tokens, emit, 1, -1);
        rpn_push_atom(revpol, nodepend, pending, tokens, emit, &atom("a"));
        rpn_push_atom(revpol, nodepend, pending, tokens, emit, &atom("b"));
    });
    println!("root_binary={root_binary}");

    let nested_invisible = render(|revpol, nodepend, pending, tokens, emit| {
        rpn_push_op(revpol, nodepend, pending, tokens, emit, 1, -1);
        rpn_push_atom(revpol, nodepend, pending, tokens, emit, &atom("a"));
        rpn_push_op(revpol, nodepend, pending, tokens, emit, 2, -1);
        rpn_push_atom(revpol, nodepend, pending, tokens, emit, &atom("b"));
        rpn_push_atom(revpol, nodepend, pending, tokens, emit, &atom("c"));
    });
    println!("nested_invisible={nested_invisible}");

    let nested_parenthesized = render(|revpol, nodepend, pending, tokens, emit| {
        rpn_push_op(revpol, nodepend, pending, tokens, emit, 2, -1);
        rpn_push_atom(revpol, nodepend, pending, tokens, emit, &atom("a"));
        rpn_push_op(revpol, nodepend, pending, tokens, emit, 1, -1);
        rpn_push_atom(revpol, nodepend, pending, tokens, emit, &atom("b"));
        rpn_push_atom(revpol, nodepend, pending, tokens, emit, &atom("c"));
    });
    println!("nested_parenthesized={nested_parenthesized}");
}
